plenum/server/replica.py

```diff
@@ -2539,86 +731,46 @@ def send(self, msg, stat=None) -> None:
         self.logger.trace("{} sending {}".format(self, msg.__class__.__name__),
                           extra={"cli": True, "tags": ['sending']})
         self.logger.trace("{} sending {}".format(self, msg))
-        if stat:
-            self.stats.inc(stat)
+        if to_nodes:
+            self.node.sendToNodes(msg, names=to_nodes)
+            return
         self.outBox.append(msg)
 
     def revert_unordered_batches(self):
         """
         Revert changes to ledger (uncommitted) and state made by any requests
         that have not been ordered.
         """
-        i = 0
-        for key in sorted(self.batches.keys(), reverse=True):
-            if compare_3PC_keys(self.last_ordered_3pc, key) > 0:
-                ledger_id, discarded, _, prevStateRoot, len_reqIdr = self.batches.pop(key)
-                discarded = invalid_index_serializer.deserialize(discarded)
-                self.logger.debug('{} reverting 3PC key {}'.format(self, key))
-                self.revert(ledger_id, prevStateRoot, len_reqIdr - len(discarded))
-                i += 1
-            else:
-                break
-        return i
+        return self._ordering_service.revert_unordered_batches()
 
     def on_catch_up_finished(self, last_caught_up_3PC=None, master_last_ordered_3PC=None):
-
         if master_last_ordered_3PC and last_caught_up_3PC and \
                 compare_3PC_keys(master_last_ordered_3PC,
                                  last_caught_up_3PC) > 0:
             if self.isMaster:
                 self._caught_up_till_3pc(last_caught_up_3PC)
             else:
+                self._ordering_service.first_batch_after_catchup = True
                 self._catchup_clear_for_backup()
-                self.first_batch_after_catchup = True
-        self.stasher.unstash_catchup()
+        self.stasher.process_all_stashed(STASH_CATCH_UP)
+
+    def discard_req_key(self, ledger_id, req_key):
```

Comment:
Is `discard_req_key` used anywhere?

Reply:
This is used in the Node in the method _clear_request_for_txn()